File: Week_03/generateParenthesis.go
func generateParenthesis(n int) []string {
	ret :=[]string{}
	var generate func(left, right int,str string)
	generate=func(left ,right int,str string){
			if n == left && n == right {
					ret = append(ret,str)
					return 
			}
			if left < n{
					generate(left+1,right,str+"(")
			}
			if right < left{
					generate(left,right+1,str+")")
			}
	}
	generate(0,0,"")
	return ret
}

File: Week_07/generateParenthesis.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
    let ret = [];
    let dfs = (str, left, right) => {
        if (left == n && right == n) {
            return ret.push(str);
        }
        if (left < n) dfs(str + "(", left + 1, right);
        if (right < left) dfs(str + ")", left, right + 1);
    };
    dfs("", 0, 0);
    return ret;
};
